File: api/src/main/java/org/apache/flink/agents/api/Event.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.flink.agents.api;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.core.JsonParser;
import com.fasterxml.jackson.databind.DeserializationContext;
import com.fasterxml.jackson.databind.JsonDeserializer;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import org.apache.flink.agents.api.context.MemoryRef;

import java.io.IOException;
import java.util.HashMap;
import java.util.Map;
import java.util.Objects;
import java.util.UUID;

/** Base class for all event types in the system. */
public class Event {

    private static final ObjectMapper MAPPER = new ObjectMapper();

    private final UUID id;
    private final String type;
    private final Map<String, Object> attributes;

    // Keep the annotation on the field as well as the creator parameter so it also applies when
    // Jackson constructs Event subclasses whose creators do not declare attachments.
    @JsonDeserialize(contentUsing = AttachmentValueDeserializer.class)
    private final Map<String, Object> attachments;

    /**
     * Runtime-internal timestamp from the source record. Not part of the cross-language event
     * contract; used by the Flink runtime for timestamp propagation.
     */
    private Long sourceTimestamp;

    /** Unified event with user-defined type and attributes. */
    public Event(String type, Map<String, Object> attributes) {
        this(UUID.randomUUID(), type, attributes, new HashMap<>());
    }

    /** Unified event with user-defined type and empty attributes. */
    public Event(String type) {
        this(type, new HashMap<>());
    }

    public Event(
            @JsonProperty("id") UUID id,
            @JsonProperty("type") String type,
            @JsonProperty("attributes") Map<String, Object> attributes) {
        this(id, type, attributes, new HashMap<>());
    }

    @JsonCreator
    public Event(
            @JsonProperty("id") UUID id,
            @JsonProperty("type") String type,
            @JsonProperty("attributes") Map<String, Object> attributes,
            @JsonProperty("attachments")
                    @JsonDeserialize(contentUsing = AttachmentValueDeserializer.class)
                    Map<String, Object> attachments) {
        if (type == null || type.isEmpty()) {
            throw new IllegalArgumentException("Event 'type' must not be null or empty.");
        }
        this.id = id;
        this.type = type;
        this.attributes = attributes != null ? new HashMap<>(attributes) : new HashMap<>();
        this.attachments = attachments != null ? new HashMap<>(attachments) : new HashMap<>();
    }

    public UUID getId() {
        return id;
    }

    /** Returns the event type string used for routing. */
    @JsonProperty("type")
    public String getType() {
        return type;
    }

    public Map<String, Object> getAttributes() {
        return attributes;
    }

    public Map<String, Object> getAttachments() {
        return attachments;
    }

    public Object getAttr(String name) {
        return attributes.get(name);
    }

    public void setAttr(String name, Object value) {
        attributes.put(name, value);
    }

    public Object getAttachment(String name) {
        return attachments.get(name);
    }

    public void setAttachment(String name, Object value) {
        attachments.put(name, value);
    }

    @JsonIgnore
    public boolean hasSourceTimestamp() {
        return sourceTimestamp != null;
    }

    @JsonIgnore
    public Long getSourceTimestamp() {
        return sourceTimestamp;
    }

    @JsonIgnore
    public void setSourceTimestamp(long timestamp) {
        this.sourceTimestamp = timestamp;
    }

    /**
     * Creates a base Event from another Event, copying id, type, attributes, and attachments.
     * Subclasses override this to reconstruct typed event objects with proper field
     * deserialization.
     */
    public static Event fromEvent(Event event) {
        Event copy =
                new Event(
                        event.getId(),
                        event.getType(),
                        new HashMap<>(event.getAttributes()),
                        new HashMap<>(event.attachments));
        if (event.hasSourceTimestamp()) {
            copy.setSourceTimestamp(event.getSourceTimestamp());
        }
        return copy;
    }

    /**
     * Creates an Event from a JSON string.
     *
     * @param json the JSON string to deserialize
     * @return the deserialized Event
     * @throws IOException if JSON parsing fails or the 'type' field is missing or empty
     */
    public static Event fromJson(String json) throws IOException {
        return MAPPER.readValue(json, Event.class);
    }

    /** Deserializes one attachment value, preserving explicitly tagged memory references. */
    static final class AttachmentValueDeserializer extends JsonDeserializer<Object> {

        @Override
        public Object deserialize(JsonParser parser, DeserializationContext context)
                throws IOException {
            JsonNode node = parser.getCodec().readTree(parser);
            if (node.isObject()
                    && MemoryRef.TYPE_VALUE.equals(node.path(MemoryRef.TYPE_FIELD).asText())) {
                return parser.getCodec().treeToValue(node, MemoryRef.class);
            }
            return parser.getCodec().treeToValue(node, Object.class);
        }
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Event other = (Event) o;
        return Objects.equals(this.id, other.id)
                && Objects.equals(this.getType(), other.getType())
                && Objects.equals(this.attributes, other.attributes)
                && Objects.equals(this.attachments, other.attachments);
    }

    @Override
    public int hashCode() {
        return Objects.hash(id, getType(), attributes, attachments);
    }
}

File: python/flink_agents/api/events/event.py
################################################################################
#  Licensed to the Apache Software Foundation (ASF) under one
#  or more contributor license agreements.  See the NOTICE file
#  distributed with this work for additional information
#  regarding copyright ownership.  The ASF licenses this file
#  to you under the Apache License, Version 2.0 (the
#  "License"); you may not use this file except in compliance
#  with the License.  You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
# limitations under the License.
#################################################################################
import hashlib
import json
from typing import Any, Dict

try:
    from typing import override
except ImportError:
    from typing_extensions import override
from uuid import UUID

from pydantic import BaseModel, Field, model_validator
from pydantic_core import PydanticSerializationError
from pyflink.common import Row


class Event(BaseModel, extra="allow"):
    """Base class for all event types in the system.

    This class serves dual purposes:

    - **Unified events**: Instantiated directly with a user-defined ``type``
      string and arbitrary key-value ``attributes``.  No subclassing required.
    - **Subclassed events**: Traditional usage where concrete subclasses (e.g.,
      :class:`InputEvent`) extend this class.  The ``type`` defaults to the
      fully qualified class name.

    Event allows extra properties, but these must be BaseModel instances or JSON
    serializable.

    Attributes:
    ----------
    id : UUID
        Unique identifier for the event, generated deterministically based on
        event content.
    type : str | None
        User-defined event type for routing.  When *None*, :meth:`get_type`
        falls back to the fully qualified class name.
    attributes : Dict[str, Any]
        Arbitrary key-value properties for unified events.
    """

    id: UUID = Field(default=None)
    type: str | None = Field(default=None)
    attributes: Dict[str, Any] = Field(default_factory=dict)

    @staticmethod
    def __serialize_unknown(field: Any) -> Dict[str, Any]:
        """Handle serialization of unknown types, specifically Row objects."""
        if isinstance(field, Row):
            return {"type": "Row", "values": field._values}
        else:
            err_msg = f"Unable to serialize unknown type: {field.__class__}"
            raise PydanticSerializationError(err_msg)

    @override
    def model_dump_json(self, **kwargs: Any) -> str:
        """Override model_dump_json to handle Row objects using fallback."""
        # Set fallback if not provided in kwargs
        if "fallback" not in kwargs:
            kwargs["fallback"] = self.__serialize_unknown
        return super().model_dump_json(**kwargs)

    def _generate_content_based_id(self) -> UUID:
        """Generate a deterministic UUID based on event content using MD5 hash.

        Similar to Java's UUID.nameUUIDFromBytes(), uses MD5 for version 3 UUID.
        """
        # Serialize content excluding 'id' to avoid circular dependency
        content_json = super().model_dump_json(
            exclude={"id"}, fallback=self.__serialize_unknown
        )
        md5_hash = hashlib.md5(content_json.encode()).digest()
        return UUID(bytes=md5_hash, version=3)

    @model_validator(mode="after")
    def validate_and_set_id(self) -> "Event":
        """Validate that fields are serializable and generate content-based ID."""
        if self.id is None:
            object.__setattr__(self, "id", self._generate_content_based_id())
        self.model_dump_json()
        return self

    def __setattr__(self, name: str, value: Any) -> None:
        super().__setattr__(name, value)
        # Ensure added property can be serialized.
        self.model_dump_json()
        # Regenerate ID if content changed (but not if setting 'id' itself)
        if name != "id":
            object.__setattr__(self, "id", self._generate_content_based_id())

    def get_type(self) -> str:
        """Return the event type used for routing.

        For unified events (where ``type`` is explicitly set), returns the
        user-defined type string.  For subclasses, defaults to the fully
        qualified class name (``module.ClassName``).
        """
        if self.type is not None:
            return self.type
        return f"{self.__class__.__module__}.{self.__class__.__qualname__}"

    def get_attr(self, name: str) -> Any:
        """Get an attribute value from the attributes map."""
        return self.attributes.get(name)

    def set_attr(self, name: str, value: Any) -> None:
        """Set an attribute value in the attributes map."""
        self.attributes[name] = value

    @classmethod
    def from_json(cls, json_str: str) -> "Event":
        """Deserialize a unified event from a JSON string.

        Parameters
        ----------
        json_str : str
            JSON string containing at least a ``type`` field.

        Returns
        -------
        Event
            The deserialized event.

        Raises
        ------
        ValueError
            If the ``type`` field is missing or empty.
        """
        data = json.loads(json_str)
        event = cls.model_validate(data)
        if event.type is None or event.type == "":
            msg = "Event JSON must contain a 'type' field."
            raise ValueError(msg)
        return event


class InputEvent(Event):
    """Event generated by the framework, carrying an input data that
    arrives at the agent.
    """

    input: Any


class OutputEvent(Event):
    """Event representing a result from agent. By generating an OutputEvent,
    actions can emit output data.

    Attributes:
    ----------
    output : Any
        The output result returned by the agent.
    """

    output: Any
